Observation Noise

adam2392 opened the pull request:
Added observation noise for the Projections class.

Marmaduke, I switched some implementations from our original convo. Let me know if they are incorrect.

Note:
- I switched to the projections class because it did not make sense how to add noise to the Monitors class, if the monitors for the source signals would already have noise. 
- some issues I am not sure of on regarding the dt, shape, nsig terms in setting the observation noise, and configuration.

maedoc committed:
add noise to base monitor

adam2392 committed:
OBSERVATION NOISE

Adding observation noise. Can be used as a keyword argument into
any Projection subclass.

For example, user sets an obsnoise object from the
Additive, or Multiplicative class:
            mon_SEEG = monitors.iEEG.from_file(period=period,
                                           variables_of_interest=[1],
                                           obsnoise=obsnoise)

To Do:
1. Check with TVB member that this works as intended. Works rn
on my local machine.
2. Does not generate line noise. Needs to be generated at a specific
frequency bands.
3. Needs to check dt, shape terms within Projection sample function.
4. How should nsig be determined. Should it have size (1,), or (numchans,)
since we are adding noise to the SEEG channels?

maedoc commented:
hi @adam2392 is there a related notebook or demo for this? if not, consider adding one to the [docs demo folder](https://github.com/the-virtual-brain/tvb-documentation/tree/master/demos), and link to it here.

adam2392 commented:
Created a pull request within the tvb-documentation. The corresponding doc should be at this directory:

https://github.com/the-virtual-brain/tvb-documentation/tree/master/demos

Let me know if there is anything else needed here.

maedoc commented on tvb/simulator/monitors.py:
```diff
@@ -514,6 +521,21 @@ def sample(self, step, state):
+                if self.obsnoise.ntau > 0.0:
+                    self.obsnoise.configure_coloured(self.dt, sample.shape)
+                else:
+                    self.obsnoise.configure_white(self.dt)
```
configuration should be done in the monitor's configure step if possible.  

> Should dt be the same as in the source level?

yes

> how should shape be set here for colored noise?

looks correct to me

adam2392 committed:
FIXED COMMENTS FOR CONFIGURATION

Moved noise configuration under the configure step.
Then tested the shape and simulation with additive/coloured noise.

maedoc commented:
@adam2392 if you need to pull this contrib also please close the PR

adam2392 commented:
I don't need to close this I believe. This was just an additional few lines of code to allow user to add noise generated post-simulation.

maedoc commented:
I think this is still a good addition, so I'm going to merge this.  Thanks @adam2392
